### ISSUE TITLE
Using DbContext with Simple Injector for Integration Test v Runtime Configuration

### ISSUE THREAD

**Issue by andez2000** [simpleinjector/SimpleInjector#609]
Are there any examples, tips, or recommendations of working with InMemory `DbContext` for Integration Tests with ASPNET Core WebApi whilst using Simple Injector?  

Microsoft added an example for this on GitHub which I tried to follow.  See [https://docs.microsoft.com/en-us/ef/core/miscellaneous/testing/in-memory](https://docs.microsoft.com/en-us/ef/core/miscellaneous/testing/in-memory)

The only thing you can access back in the `WebApplicationFactory` is the configured services - not strictly the Simple Injector container without adding that into the Service Collection from StartUp.  So the initialisation of the DbContext needs to be done from the test project and find its way to the Simple Injector container.  And with this, this seems the only viable way to add the test data to the DbContext.

Additionally [https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-2.1](https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-2.1) talks about using `WebApplicationFactory` with entity framework.

**Comment by andrebugs2** [comment]
I'm in a similar situation. I have integration tests that test my REST API using the WebApplicationFactory method described at https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-2.1. I have a custom WebApplicationFactory where I set up the DBContext to use a special integration test database and seed the database. I can access the DBContext by looking it up in builder.ConfigureServices().

I am now in the process of switching to Simple Injector. My DBContext is only registered in Simple Injector (I've seen you recommend that in other threads, that only framework-related things should be registered in the built-in DI container). How can I seed the DB in WebApplicationFactory when I cannot access the Simple Injector container? I can set the environment to "IntegrationTesting" and look at that in Startup.cs I guess, and in that case seed the DB. My custom WebApplicationFactory is in my test project and it would feel cleaner to do the DB seeding there.

**Comment by dotnetjunkie** [comment]
Hi @andez2000,

If you are running your integration tests in a different process (or even a different AppDomain) from the process where you are running Simple Injector, there is no way to influence the Container. This isn't specific to Simple Injector, but it works the same for all DI Containers.

If you look closely at that ASP.NET Core integration test documentation, however, you'll see that they are actually *not* running the application out-of-process, but loading everything in the same process. Although I'm not experienced with the `WebApplicationFactory<T>` and unfortunately don't have time to go through that complete article, integrating Simple Injector would probably be very similar.

**Comment by andrebugs2** [comment]
I realized that it probably doesn't really matter how I access the DBContext just for seeding the database, the end result is the same. The application itself will use Simple Injector during the test to access the DBContext.

**Comment by jeffrey-igaw** [comment]
@andrebugs2 @dotnetjunkie Yeap, your the end result is right. Whatever we registered services add DbContext or other fake services, we can't still resolve this issue when we using SimpleInjector.

because, service registration process is below in integrated tests

1. `WebApplicationFactory<Startup>.ConfigureWebHost`
2. `Facotry.WithWebHostBuilder()`
3. `Startup.ConfigureServices`
4. Run Test

So we doesn't have any opportunaty for overwriting registration.
Actually this is not issue about simpleinjector, this is **issue of asp.net core project**.

But.. this issue can be resolved with simple and dirty way.

1. Container as public static field of `Startup.cs` and use this.  
  What the..temptation of the Evil?
2. Don't use SimpleInjector like third party DI Container  
  Wow this is reaally awesome. we are Romeo and Juliet?
3. Use environment value(`ASPNETCORE_ENV`?) and register DbContext using InMemory provider when testing.  
  This is a really no worthyless way, our builds result will contain in memory provider.

I love to use simpleinjector and this saved me from legacy test hell with contains necessary features in real world and..... **FAST**.

**Comment by dotnetjunkie** [comment]
I don't consider it to be evil if you make the Container public and accessible from your test suite.

**Comment by jeffrey-igaw** [comment]
In my experience, if container instance is public, sometimes someone or I will want to use in some static method as `Startup.Container.GetInstance<IService>`... :cry: and process will be broken cause of mis-scope and long live instance need more resources. I would like to prevent it structurally. 

I will research more, Thank you. 👍

**Comment by dotnetjunkie** [comment]
You should always prevent using the container as Service Locator. So care must be taken.

**Comment by voroninp** [comment]
@jeffrey-igaw 
> 1. A Container as public **static** field of `Startup.cs` and use this.

Unless you want to mock services differently, and run tests in parallel.

**Comment by jagorath** [comment]
I resolved the issue this way in netcore 2.2 by making the SimpleInjector container a public property of the Startup class, creating a wrapper class/interface for it, and registering this class with the built-in Asp.Net core DI.  The container can then be resolved in the integration test.

Kind of hacky but at least no static fields etc.

```
    public interface ISimpleInjectorContainer
    {
        Container TheContainer { get; }
    }

    public class SimpleInjectorContainer: ISimpleInjectorContainer
    {
        public SimpleInjectorContainer(Container theContainer)
        {
            TheContainer = theContainer;
        }

        public Container TheContainer { get; }
    }

    public class Startup
    {
        protected readonly Container _container = new Container();
etc...

        public void ConfigureServices(IServiceCollection services)
            services.AddSingleton<ISimpleInjectorContainer>(new SimpleInjectorContainer(_container));
etc..

In the unit test:
                var factory = CreateFactory();
                var client = factory.CreateClient();
                var host = factory.Server?.Host

                using (var scope = host.Services.CreateScope())
                {
                    var services = scope.ServiceProvider;
                    var container = services.GetRequiredService<ISimpleInjectorContainer>().TheContainer;
                    var queryHandler = container.GetInstance<IQueryProcessor>();

```

Hope this helps.
